paf-mvp-cmp/src/controller.ts

```diff
@@ -0,0 +1,565 @@
+import { Locale } from './locale';
+import { Config } from './config';
+import { BindingShowRandomId, BindingThisSiteOnly } from './bindings';
+import { Log } from '@core/log';
+import { BindingButton, BindingChecked, BindingCheckedMap, BindingElement, BindingViewOnly } from '@core/ui/binding';
+import { Identifier, IdsAndOptionalPreferences, Preferences, PreferencesData } from '@core/model/generated-model';
+import {
+  getIdsAndPreferences,
+  refreshIdsAndPreferences,
+  signPreferences,
+  getNewId,
+  saveCookieValue,
+  updateIdsAndPreferences,
+  removeCookie,
+  ShowPromptOption,
+} from '@frontend/lib/paf-lib';
+import { Marketing, Model } from './model';
+import { PafStatus } from '@core/operator-client-commons';
+import { View } from './view';
+import { getCookieValue } from '@frontend/utils/cookie';
+import { Cookies, getPrebidDataCacheExpiration } from '@core/cookies';
+import { TcfCore } from './tcfcore';
+
+/**
+ * Controller class used with the model and views. Uses paf-lib for data access services.
+ */
+export class Controller {
+  // The model the controller is manipulating.
+  private readonly model = new Model();
+
+  // The locale that the UI should adopt.
+  private readonly locale: Locale;
+
+  // The options provided to the controller.
+  private readonly config: Config;
+
+  // The view associated with the controller.
+  private readonly view: View;
+
+  // Timer used to hide the snackbar.
+  private countDown: NodeJS.Timer;
+
+  private log: Log;
+
+  /**
+   * Constructs a new instance of Controller.
+   * @param script element this method is contained within
+   * @param locale the language file to use with the UI
+   * @param config the configuration for the controller
+   */
+  constructor(script: HTMLOrSVGScriptElement, locale: Locale, config: Config, log: Log) {
+    this.locale = locale;
+    this.config = config;
+    this.log = log;
+    this.view = new View(script, locale, config);
+    this.model.onlyThisSiteEnabled = config.siteOnlyEnabled;
+    this.mapFieldsToUI(); // Create the relationship between the model fields and the UI elements
+    this.load()
+      .then(() => this.display())
+      .catch((e) => log.Error('constructor', e));
+  }
+
+  /**
+   * Displays the card most appropriate given the current state of the data model.
+   * @remarks
+   * If all the data is persisted then show the snackbar.
+   * If none of the data is persisted then show the intro card or the settings depending on configuration.
+   * If some of the data is persisted and others not then show the settings card.
+   */
+  public async display(card?: string) {
+    if (card === null || card === undefined) {
```

Comment:
- don't think it needs to be async
- Consider refactoring this extract a `getCard()` method that would do quick returns instead of these nested tests, I think it would be much more readable.  Something like:
```typescript
public async display(card?: string) {
        this.setCard(card ?? this.getCard());
}

// ⚠️ not tested
private getCard(): string|undefined {
        if (this.model.status === PafStatus.NOT_PARTICIPATING) {
            return undefined;
        }
        if (this.model.allPersisted && this.model.status === PafStatus.PARTICIPATING) {
            return 'snackbar';
        }

        if (this.model.nonePersisted && this.config.displayIntro && this.model.status === PafStatus.REDIRECT_NEEDED) {
            return 'intro';
        }
        if (this.model.status !== PafStatus.REDIRECT_NEEDED) {
            return 'settings';
        }
    }

```

Reply:
Will be changed in the next push.